auth/providers/azure/pop_tokenverifier_test.go

```diff
@@ -143,160 +144,330 @@ const (
+func NewPoPTokenBuilder() *PoPTokenBuilderImpl {
+	return &PoPTokenBuilderImpl{}
+}
+
+func (b *PoPTokenBuilderImpl) SetTimestamp(ts int64) *PoPTokenBuilderImpl {
+	b.ts = ts
+	return b
+}
+
+func (b *PoPTokenBuilderImpl) SetKid(kid string) *PoPTokenBuilderImpl {
+	b.kid = kid
+	return b
+}
+
+func (b *PoPTokenBuilderImpl) SetHostName(hostName string) *PoPTokenBuilderImpl {
+	b.hostName = hostName
+	return b
+}
+
+// A method that sets the header of the PoP token
+func (b *PoPTokenBuilderImpl) SetHeader() error {
+	keyID := b.popKey.KeyID()
+	algo := b.popKey.Alg()
+	typ := "pop"
+	if b.kid == headerBadKeyID {
+		keyID = ""
 	}
-	popKey, err := NewSWPoPKey()
-	if err != nil {
-		return "", fmt.Errorf("Failed to generate Pop key. Error:%+v", err)
+	if b.kid == headerBadAlgo {
+		algo = "wrong"
+	}
+	if b.kid == headerBadtyp {
+		typ = "wrong"
 	}
+	header := fmt.Sprintf(`{"typ":"%s","alg":"%s","kid":"%s"}`, typ, algo, keyID)
 
-	key, err := NewSwkKey()
-	if err != nil {
-		return "", fmt.Errorf("Failed to generate SF key. Error:%+v", err)
+	if b.kid == headerBadtypType {
+		wrongTyp := 1
+		header = fmt.Sprintf(`{"typ":%d,"alg":"%s","kid":"%s"}`, wrongTyp, algo, keyID)
+	}
+	if b.kid == headerBadtypMissing {
+		header = fmt.Sprintf(`{"alg":"%s","kid":"%s"}`, algo, keyID)
 	}
+	b.token.Header = base64.RawURLEncoding.EncodeToString([]byte(header))
+	return nil
+}
 
+// A method that sets the payload of the PoP token
+func (b *PoPTokenBuilderImpl) SetPayload() error {
 	var cnf string
-	if kid == atCnfClaimWrong {
+	if b.kid == atCnfClaimWrong {
 		cnf = "wrongCnf"
 	} else {
-		cnf = popKey.KeyID()
+		cnf = b.popKey.KeyID()
 	}
 
-	var accessTokenData string
-	if kid == atCnfClaimMissing {
+	nonce := uuid.New().String()
+	nonce = strings.Replace(nonce, "-", "", -1)
+
+	accessTokenData := fmt.Sprintf(popAccessToken, time.Now().Add(time.Minute*5).Unix(), cnf)
+	if b.kid == atCnfClaimMissing {
 		accessTokenData = fmt.Sprintf(popAccessTokenWithoutCnf, time.Now().Add(time.Minute*5).Unix())
-	} else {
-		accessTokenData = fmt.Sprintf(popAccessToken, time.Now().Add(time.Minute*5).Unix(), cnf)
+	}
+	if b.kid == accessTokenCnfWrong {
+		accessTokenData = fmt.Sprintf(`{ "aud": "client", "iss" : "kd", "exp" : "%d","cnf": {"kid":1} }`, time.Now().Add(time.Minute*5).Unix())
 	}
 
-	at, err := key.GenerateToken([]byte(accessTokenData))
+	at, err := b.swkKey.GenerateToken([]byte(accessTokenData))
 	if err != nil {
-		return "", fmt.Errorf("Error when generating token. Error:%+v", err)
+		return fmt.Errorf("Error when generating token. Error:%+v", err)
 	}
-	var header, headerB64 string
-	nonce := uuid.New().String()
-	nonce = strings.Replace(nonce, "-", "", -1)
-	keyID := popKey.KeyID()
-	algo := popKey.Alg()
-	typ := "pop"
-	if kid == headerBadKeyID {
-		keyID = ""
+
+	payload := fmt.Sprintf(`{ "at" : "%s", "ts" : %d, "u": "%s", "cnf":{"jwk":%s}, "nonce":"%s"}`, at, b.ts, b.hostName, b.popKey.Jwk(), nonce)
```

Comment:
Can this be a struct as well ?

Reply:
If we change this to a struct, then type of each of the fields will be fixed. But, some of the test cases are to identify errors when the wrong type of argument is passed. If we make it a struct then we won't be able to test those. For example: the right type of `ts` is int64, but in `TsClaimsTypeString` case, we pass `ts` as a string.